### PR TITLE
 Design: Clustering in SQLflow

### PR DESCRIPTION
**Pull request by Echo9573**
#648

### CODE REVIEW

File: doc/cluster_design.md [hunk 1]
```diff
@@ -0,0 +1,61 @@
+# Design: Clustering in SQLflow to analyze patterns in data
+
+## Concept
+
+For analysts and real business people, in the daily analysis work, most of the work is not prediction, but analysis of the patterns in the data. This can help them mine user behavioral characteristics and differences, helping the business discover value and operate.
+
+This design doc introduces how to support the `Cluster Model` in SQLFlow.
```
Comment (Yancey0623) [inline doc/cluster_design.md]: Add a section to introduce the `Cluster Model`?

File: doc/cluster_design.md [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+## User interface
+
+Users usually use a **TRAIN SQL** to train a model in Supervised learning. But, in this scenario, we focus on the extraction of data patterns in unsupervised learning. Therefore, we use **EXTRCT SQL** for pattern extraction, the simple pipeline like:
+
+EXTRCT SQL:
+
+``` sql
+SELECT * FROM train_table
+EXTRCT clusterModel
```
Comment (Yancey0623) [inline doc/cluster_design.md]: How do you want to expose `ClusterModel`  to SQLFlow? For the current implement this should be a [Tensorflow premade Estimator](https://www.tensorflow.org/guide/premade_estimators) or [Custom Kerse Model](https://keras.io/models/about-keras-models/#model-subclassing), can we implement `ClusterModel` as a Custom Kerase Model?

File: doc/cluster_design.md [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+- `new_table`: If you want to apply the model that extracts from train_table to the new data new_table directly, you can use **PREDICT SQL**. Note that the structure of new_table is the same as train_table, namely, same feature column.
+- `result_test_table` is the cluster result for new_table.
+
+## Implement Details
+-
+- 
+- 
+
+## Note
+- The **EXTRCT SQL** includes two models, the autoencode model and the cluster model.
```
Comment (Yancey0623) [inline doc/cluster_design.md]: Maybe we need a more general target for the keyword **EXTRACT**

Comment (Yancey0623) [inline doc/cluster_design.md]: `EXTRCT SQL ` => `EXTRACT SQL`

Comment (Echo9573) [reply doc/cluster_design.md]: This keyword is just for illustration of the design. Next, maybe we can think about a better keywords together.

File: doc/cluster_design.md [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+## Concept
+
+For analysts and real business people, in the daily analysis work, most of the work is not prediction, but analysis of the patterns in the data. This can help them mine user behavioral characteristics and differences, helping the business discover value and operate.
+
+This design doc introduces how to support the `Cluster Model` in SQLFlow. 
+
+## User interface
+
+Users usually use a **TRAIN SQL** to train a model in Supervised learning. But, in this scenario, we focus on the extraction of data patterns in unsupervised learning. Therefore, we use **EXTRCT SQL** for pattern extraction, the simple pipeline like:
```
Comment (terrytangyuan) [inline doc/cluster_design.md]: Do you mean EXTRACT?

**Comment by terrytangyuan** [comment]
Are we planning to support semi-supervised clustering? If so, we may allow users to optionally pass a subset of labels.

File: doc/cluster_design.md (hunk 4 above)
Comment (Echo9573) [reply doc/cluster_design.md]: yes~

File: doc/cluster_design.md (hunk 2 above)
Comment (typhoonzero) [inline doc/cluster_design.md]: "EXTRACT" a model does not read well, we can train a model, save a model, use a model to do prediction but not extract a model. Since the model name is "clusterModel", we can distinguish the model type just like we did in xgboost. Maybe we can use `TRAIN ClusterModel.myDeepClusteringModel` then the code generator can generate specific code for the training pipeline for clustering models.

File: doc/cluster_design.md [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+- 
+- 
+
+## Note
+- The **EXTRCT SQL** includes two models, the autoencode model and the cluster model. 
+First, the former is used to achieve data compression. At training time, the input to this model is train_table (eg. train_table.shape = (10000 * 184)), and the output is also train_table. We only use the output of the trained encode layer (10000*7) as the input to the clustering model.
+Then, the clustering model starts training, randomly initializes weights and multiple iterations, generates clustering models and clustering results of train_table.
+Next, we average each feature in the train_table according to the clustering result category to get the result table.
+Finally, **EXTRCT SQL** outputs two parts, clustering model and the result table are included.
+- The example of result_table:
```
Comment (tonyyang-svail) [inline doc/cluster_design.md:137]: How would an analyst use this table?

**Review by tonyyang-svail — COMMENTED** [review]
Hi @Echo9573, thanks for this informative design doc. It helped the team to realize a significant portion of an analysts work is unsupervised clustering.

File: doc/cluster_design.md (hunk 5 above)
Comment (Echo9573) [reply doc/cluster_design.md:137]: Analysts will compare the mean of each feature to analyze the behavioral characteristics and differences of each group of users, maybe by ploting the result_table.

**Comment by Echo9573** [comment]
> Are we planning to support semi-supervised clustering? If so, we may allow users to optionally pass a subset of labels.

We would not consider semi-supervised learning at present, because it may not be widely used among analysts.

**Comment by terrytangyuan** [comment]
> > Are we planning to support semi-supervised clustering? If so, we may allow users to optionally pass a subset of labels.
> 
> We would not consider semi-supervised learning at present, because it may not be widely used among analysts.

Is there a reference for that? It is pretty popular in applications where there are not sufficient labels or that the labels are polluted. This will help us decide whether we need to reuse `TRAIN` for unsupervised problems and `LABEL` for semi-supervised problems instead of inventing a new `EXTRACT` syntax.

**Comment by Yancey0623** [comment]
> Is there a reference for that? It is pretty popular in applications where there are not sufficient labels or that the labels are polluted. This will help us decide whether we need to reuse TRAIN for unsupervised problems and LABEL for semi-supervised problems instead of inventing a new EXTRACT syntax.

@terrytangyuan @Echo9573 

From the design doc, maybe the root reason for the `EXTRACT` keyword is: 
* The output of ClusteringModel including **trained model** and the **result of clustering** which difference from TRAIN SQL.
* Other supervised algorthms like KMeans only need to output the result, don't need to train a model.

Maybe we can reuse the `TRAIN` keyword and extend `INTO statement` to :

``` sql
SELEC * FROM train_table
TRAIN ClusteringModel
INTO
    model = my_cluster_model
    table = cluster.predict
``` 

For the KMeans algorithms:

``` sql
SELECT * FROM train_table
TRAIN KMeans
INTO
    table = kmeans.predict
```

File: doc/cluster_design.md (hunk 5 above)
Comment (tonyyang-svail) [reply doc/cluster_design.md:137]: I see. Thanks for the explanation. :)

File: doc/cluster_design.md [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+- `input_table` is the high-dimensional table to be clustered.
+- `model.encode_units` is the autoencoder model layer's encoder units, the decode_units can reverse encode_units directly.
+- `model.n_clusters` is the number of patterns after clustering.
+- `my_cluster_model` is the trained cluster model.
+- `run_pretrain`  is used to determine if autoencoder pretrain needs to be run, default true.
+- `model.existed_pretrain_model` is used to specify an existing pretrain_model
+- `output_table` is the cluster result for input_table, which is adding the `group_id` column predicted by the cluster model to the input_table.
+
+## clusterModel Details
+<img src="figures/cluster_model_train_overview.png">
```
Comment (typhoonzero) [inline doc/cluster_design.md]: Maybe paste the link here before this figure, so that readers can understand this figure well.

File: doc/cluster_design.md [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+3. Finally, the overall train process ultimately outputs an unsupervised clustering model.
+
+
+## Implement Details
+- sqlflow_models/clusterModel.py
+
+```python
+class  clusterModel(tf.keras.Model):
+
+	def pre_train(dataset):
```
Comment (typhoonzero) [inline doc/cluster_design.md]: please remove the blank lines and use 4 spaces as indentation.

File: doc/cluster_design.md [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+## Note
+
+- The user can choose whether to run pre_train before the cluster model, ie run_pretrain=true. The user can also choose to load the already trained model by loading the existed_pretrain_model.
+Therefore, there are four cases in total:
+1.  run_pretrain = true & Using model.existed_pretrain_model = None：
+Autoencoder Pretrain + Random initialization weights for cluster. (Note that model.encode_units `is work` at this time.)
+2.  run_pretrain = true & Using model.existed_pretrain_model = existed_pretrain_model：
+existed_pretrain_model Pretrain+ Random initialization weights for cluster. (Note that model.encode_units `is not work` at this time.)
+3.  run_pretrain = false & Using model.existed_pretrain_model = None: 
+Random initialization weights for cluster. (Note that model.encode_units `is not work` at this time.)
```
Comment (terrytangyuan) [inline doc/cluster_design.md]: You mean “does not work”? Same in other places. Also did you mean to use double quotes instead?

File: doc/cluster_design.md (hunk 7 above)
Comment (Echo9573) [reply doc/cluster_design.md]: OK~~thx~

File: doc/cluster_design.md (hunk 8 above)
Comment (Echo9573) [reply doc/cluster_design.md]: The `USING` statement have a higher precedence than  pre_train=True in WITH statement. Since model.encode_units sets the pre_train part of the autoencoder network, the pre_train statement does not work when the using statement exists, so encode_units does not work.

Comment (terrytangyuan) [reply doc/cluster_design.md]: I understand that. I was just asking you to revisit this part of the grammar. It should be “does” instead of “is”. And then switch to use double quotes.

File: doc/cluster_design.md [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+
+``` sql
+SELECT * FROM input_table
+TRAIN clusterModel
+WITH
+    model.encode_units = [100, 7]
+    model.n_clusters = 5
+    model.run_pretrain = false
+COLUMN m1, m2, m3, m4, m5, m6, m7, m8, m9, m10 
+USING model.existed_pretrain_model =  existed_pretrain_model
```
Comment (Yancey0623) [inline doc/cluster_design.md]: Do we need to refine the USING syntax? Or just following the existing syntax `USING existed_pretran_model`.

File: doc/cluster_design.md [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+				p = target_distribution(q)  # update the auxiliary target distribution p
+				y_pred = q.argmax(1)
+			idx = index_array[index * batch_size: min((index+1) * batch_size, x.shape[0])]
+				loss = model.train_on_batch(x=x[idx], y=p[idx])
+			 index = index + 1 if (index + 1) * batch_size <= x.shape[0] else 0
+```
+
+- template_tf.go
+```python
+if 'pre_train' is in classifier:
```
Comment (Yancey0623) [inline doc/cluster_design.md]: It's not Python syntax, maybe using 
``` python
if hasattr(classifier, 'pre_train'):
    classifier.pre_train(...)
if hasattr(classifier, 'cluster_train_loop'):
    classifier.cluster_train_loop
```

File: doc/cluster_design.md [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+```
+
+where:
+- `input_table` is the high-dimensional table to be clustered.
+- `model.encode_units` is the autoencoder model layer's encoder units, the decode_units can reverse encode_units directly.
+- `model.n_clusters` is the number of patterns after clustering.
+- `my_cluster_model` is the trained cluster model.
+- `run_pretrain`  is used to determine if autoencoder pretrain needs to be run, default true.
+- `model.existed_pretrain_model` is used to specify an existing pretrain_model
+- `output_table` is the cluster result for input_table, which is adding the `group_id` column predicted by the cluster model to the input_table.
```
Comment (Yancey0623) [inline doc/cluster_design.md]: Maybe to specify the result column by  `PREDICT output_table.group_id` is more accurate.

File: doc/cluster_design.md (hunk 6 above)
Comment (Yancey0623) [inline doc/cluster_design.md]: How about moving  the cluster model introduction section on the top of the document, the structure can be:
1. ClusterModel introcution
1. User interface in SQLFlow
1. How implement ClusterModel it in SQLFlow

File: doc/cluster_design.md [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+
+## clusterModel Details
+<img src="figures/cluster_model_train_overview.png">
+
+The below figure demonstrates overall workflow for clusterModel train. This figure includes two parts, the pretrian autoencode model and the cluster model are included.
+1. First, the former is used to train a pretrain model. The `model.encode_units` describes the layer structure of the encoder of the autoencoder network. We only use the output of the trained encode layer (10000*7) as the input to the clustering model. 
+2. Then, the clustering model starts training, randomly initializes weights and multiple iterations, generates clustering models.
+3. Finally, the overall train process ultimately outputs an unsupervised clustering model.
```
Comment (Yancey0623) [inline doc/cluster_design.md]: How about split `Cluster` section into `Train` and `Predict` so that users can know what does `TRAIN SQL` and `PREDICT SQL` do.

File: doc/cluster_design.md (hunk 1 above)
Comment (BlackPoint-CX) [reply doc/cluster_design.md]: > As one of the most powerful method  in  pattern recognition, clustering focus on finding the similarities between items per group and differences between groups. Hence, `Cluster Model` can help analysts to build such a model which can split data samples into different group according to their features automatically.   

@Echo9573 If it can help.

File: doc/cluster_design.md (hunk 9 above)
Comment (BlackPoint-CX) [reply doc/cluster_design.md]: In my opinion, it can be divided into such two scenarios:

1. User has no pre-train model (auto-encoder). 
 In this scenario the user wants the full training process which consists of `auto-encoder` and `clustering`. Thus the user should  be forbidden to use the `USING` clause, because there is no pre-train model ready for using in training process. User should define relate parameters of `auto-encoder` model  in `WITH` clause clearly, like `model.encode_units`.

2. User has pre-train model (auto-encoder).
In this case, use has at least one pre-train `auto-encoder` model already and he/she want to use it without training this part again. `USING` clause should be used for defining the path/name of the pre-train `auto-encoder` model and  in `WITH` clause user should guarantee the correct structure of pre-train model to make sure that the model data can be loaded correctly. It requires some additional checks to be performed in the background.

File: doc/cluster_design.md (hunk 6 above)
Comment (BlackPoint-CX) [inline doc/cluster_design.md]: I think the `decoder` should be include into the stage of `Pre-train`. Because the `auto-encoder` is used for building `encoder` for next training process. The `decoder` will be created at the same time. Even the `decoder` will be never be used in the future, it still should be treated as `Pre-train`(Just my opinion).

**Review by tonyyang-svail — DISMISSED** [review]
@Echo9573 @BlackPoint-CX Thanks for submitting this excellent PR. I am approving this PR because the general design looks great to me.

Please also take a look at possible readability improvements mentioned by other reviewers. :)

File: doc/cluster_design.md (hunk 8 above)
Comment (Echo9573) [reply doc/cluster_design.md]: OK~~thx！

**Review by Yancey0623 — APPROVED** [review]
Thanks for the excellent design about unsupervised learning, LGTM and can merge this PR first, and keep improving as implement.
